Use go.mod to install tools

albertvaka opened the pull request:
### What does this PR do?

Using a `tools.go` file that is not included in the build and that imports your tools is the recommended way to install development, build and test tools using go modules. This PR adopts this recommendation and removes our previous homemade bootstrap script.

I went ahead and got rid of everywhere I found where we were using packages installed in the `GOPATH` instead of through go modules. This allows us to simplify a bit the deps caching logic in Gitlab. 

To add new tools now, there are two steps: first `go get` it, then add it to `tools.go`

### Motivation

This removes a lot of custom logic :) The idea wasn't mine, see: https://github.com/DataDog/datadog-agent/pull/6321

### Additional notes

This change causes tool licenses to end up in `LICENSE-3rdparty.csv` even if they are not built into the Agent (so, their licenses shouldn't matter) because all go packages end up in the same directory. To prevent some tools from showing up there and causing confusion (specially since some of them are GPL), I'm using the exceptions list in `.wwhrd.yml` (note: `wwhrd` is the tool that we use to generate `LICENSE-3rdparty.csv`). However, I have to parse that exceptions list manually in the invoke task, since `wwhrd` only uses that list when running the `check` command and not the `list` command.

Also note that I didn't manage to compile for Android using go modules (see comment in-code about the blockers I found), even after bumping gomobile to a version which supports modules (and having to bump the Android NDK version for it to be compatible). So I left Android using `GO111MODULE=off` as it was before. 

The Android changes here depend on the separate PR that bumps the version of gomobile and the Android NDK: https://github.com/DataDog/datadog-agent/pull/7206

lebauce commented:
You may also need to change the `//go:generate go-bindata ...` line in `pkg/status/render.go` and `cmd/agent/gui/gui.go` to `//go:generate go run github.com/shuLhan/go-bindata/cmd/go-bindata ...`
Same with `pkg/trace/pb/trace.go` with `go:generate go run github.com/tinylib/msgp/msgp ...`

albertvaka commented:
Thanks for the hint ❤️ The docs I'm following [1], though, seem to imply that it's not needed to use `go run` there, so I'm a bit confused on why this isn't working for us. Do you have any idea?

[1] https://github.com/go-modules-by-example/index/blob/master/010_tools/README.md

lebauce commented:
@albertvaka I guess it's because they run `go install`, and this one use the go.mod
If we don't run it, my understanding was that we need to use "go run" to use the version specified in go.mod

olivielpeau commented:
True, as we switch to tracking tools in `tools.go`, we're not `go install`ing the tools anymore, so we need to either replace all the tools' binary invocations with `go run <tool_package>` or find a way to run `go install` on all these tools as part of `invoke deps` (can't really think of a simple way to do this that wouldn't require reflecting the list of tools in `tools.go` in a separate file or parsing `tools.go`, both of which don't sound desirable)

albertvaka commented:
Thanks team! 👍 Using `go run` seems fine to me, but I didn't realize the example did a `go install`... and I wanted to know how it was working for them :)

albertvaka commented:
Another problem to solve is the fact that now `LICENSE-3rdparty.csv` contains tools' licenses too, some or which are GPL. Although the Agent builds don't include the tools (so we don't care about their license), there's no easy way to know if a package comes from a tool or if it doesn't when generating that file. 

All the solutions I can think of involve maintaining a list of excluded packages (parsing `tools.go` wouldn't work since it won't cover the tools' dependencies), and gotten to that point I wonder if we could simply remove `LICENSE-3rdparty.csv` and rely on the `check` feature of `wwhrd` (the tool we use to generate that file) which already has an 'exceptions list' feature. 

If all we want is to make sure we don't include GPL code, using `wwhrd check` is probably a better solution than what we have (since it will enforce that, not only generate a list), but I'm not sure if there's another reason to have the `LICENSE-3rdparty.csv` file around. Do you know that @olivielpeau?

olivielpeau commented:
`LICENSE-3rdparty.csv` is an internal requirement (see internal `Datadog Open Source policy`). It's already automatically generated with `wwhrd list` at the moment.

I don't think it's that much of a problem if `LICENSE-3rdparty.csv` lists packages that we only use as tools and aren't included in agent builds, unless they're GPL since this would cause confusion. The only solution I can think of right now is to explicitly exclude these GPL packages that we only use as tools, for example in `wwhrd`'s config, and make sure they don't end up in the generated `LICENSE-3rdparty.csv` (for example by re-using `wwhrd`'s config if `wwhrd list` still lists these packages). We'll have to be careful and make sure we only exclude packages that we only use as tools.

We could enforce the licenses with `wwhrd check` but this feels like something we could do in a separate PR.

albertvaka commented:
If it's a requirement to have the file then I will have to do something to exclude the tools from the list. They are GPL and it would cause confusion as you said.

albertvaka committed:
Use go.mod to install tools

Using a tools.go file that is not included in the build is the recommended
way to install tools using go modules. This removes the need for our
previous homemade boostrap script.

To add a new tool: first "go get" it, then add it to tools.go

albertvaka commented on .gitlab/binary_build/cluster_agent.yml:
```diff
@@ -3,10 +3,8 @@
   if: $RELEASE_VERSION_7 != ""
 
 .retrieve_linux_go_deps: &retrieve_linux_go_deps |
-  mkdir -p $GOPATH/bin && tar xzf go-bin.tar.gz -C $GOPATH/bin
-  mkdir -p $GOPATH/pkg && tar xzf go-pkg.tar.gz -C $GOPATH/pkg
   mkdir vendor && tar xzf vendor.tar.gz -C vendor
-  rm -f go-bin.tar.gz go-pkg.tar.gz vendor.tar.gz
```
We don't use anything from $GOPATH anymore

albertvaka commented on .gitlab/deps_fetch.yml:
```diff
@@ -7,20 +7,11 @@
   stage: deps_fetch
   needs: []
   script:
-    # HACK: empty the build directory cache (that can come from previous runs)
-    #       to not have remainders of previous runs, which can make our deps bootstrapping logic fail.
-    # TODO: remove this once we switch to k8s runners, they won't have this problem
-    - find "$CI_BUILDS_DIR" ! -path '*DataDog/datadog-agent*' -depth  # -delete implies -depth
-    - find "$CI_BUILDS_DIR" ! -path '*DataDog/datadog-agent*' -delete || true  # Allow failure, we can't remove parent folders of datadog-agent
```
I think we are no longer affected by this, since we don't use anything from outside our repo directory now.

albertvaka commented on .gitlab/package_build/apk.yml:
```diff
@@ -13,26 +17,15 @@ agent_android_apk:
     - when: on_success
   stage: package_build
-  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/android_builder:$DATADOG_AGENT_BUILDERS
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/android_builder:v3715617-76ed3eb
```
Updates the Android NDK. We can update it once the PR on the builders repo gets merged.

albertvaka commented on tasks/android.py:
```diff
@@ -62,8 +62,23 @@ def build(
+    ctx.run('go get golang.org/x/mobile/cmd/gomobile', env=env)
+    # Pin its version to what's indicated in go.mod
+    ctx.run(
+        'VERSION=$(grep golang.org/x/mobile go.mod | rev | cut -d - -f 1 | rev); cd /go/src/golang.org/x/mobile; git checkout $VERSION'
```
This emulates our previous "bootstrap" logic, just for this one dependency and using the `go.mod` file as the source of truth for the version.

albertvaka commented on tasks/go.py:
```diff
@@ -408,10 +352,31 @@ def generate_licenses(ctx, filename='LICENSE-3rdparty.csv', verbose=False):
     print("licenses files generated")
 
 
+# FIXME: This doesn't include licenses for non-go dependencies, like the javascript libs we use for the web gui
```
Added this FIXME because I realized this is incomplete, but it's unrelated to my changes.

KSerrania reviewed (approved):
LGTM

albertvaka committed:
Use merged builder image
